### PR TITLE
workload/schemachange: avoid JSON columns in indexes for mixed version

### LINKED ISSUE #103875: roachtest: schemachange/mixed-versions failed
roachtest.schemachange/mixed-versions [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10236439?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10236439?buildTab=artifacts#/schemachange/mixed-versions) on master @ [5d944f53668fedf017e0324475c408c3f14ae32e](https://github.com/cockroachdb/cockroach/commits/5d944f53668fedf017e0324475c408c3f14ae32e):

```
test artifacts and logs in: /artifacts/schemachange/mixed-versions/run_1
(cluster.go:2059).Run: output in run_054341.917584396_n3_workload-run-schemac: ./workload run schemachange --verbose=1 --max-ops 100 --concurrency 5 {pgurl:1-4} returned: COMMAND_PROBLEM: ssh verbose log retained in ssh_054342.738728285_n3_workload-run-schemac.log: exit status 1
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*schemachange/mixed-versions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### PR DESCRIPTION
**Pull request by fqazi**
Previously, we could generate CREATE TABLE / INDEX statements with JSON index columns which could end up failing in a mixed version state. To address this, this patch detects JSON columns and expects the appropriate errors in a mixed version state.

Fixes: #103875
Release note: None

### CODE REVIEW

File: pkg/sql/create_table.go [hunk 1]
```diff
@@ -1603,7 +1603,7 @@ func NewTableDesc(
 			}
 
 			// Version gates for enabling primary keys for JSONB columns
```
Comment (chengxiong-ruan) [inline pkg/sql/create_table.go:1605]: nit: This comment would appreciate an update.

**Review by chengxiong-ruan — APPROVED** [review]

**Comment by fqazi** [comment]
@chengxiong-ruan TFTR!

**Comment by fqazi** [comment]
Hit a flaky test, retrying

bors r+

**Comment by fqazi** [comment]
Hit an unrelated test flake, retrying one more time

bors r+

**Pull request merged**